src/main/scala/org/apache/spark/sql/redis/RedisSourceRelation.scala

```diff
@@ -257,10 +263,19 @@ class RedisSourceRelation(override val sqlContext: SQLContext,
   /**
     * read rows from redis
     */
-  private def scanRows(node: RedisNode, keys: Seq[String], schema: StructType,
+  private def scanRows(node: RedisNode, keys: Seq[String], keyType: String, schema: StructType,
                        requiredColumns: Seq[String]): Seq[Row] = {
     withConnection(node.connect()) { conn =>
-      val pipelineValues = mapWithPipeline(conn, keys) { (pipeline, key) =>
+      val filteredKeys =
+        if (filterKeysByTypeEnabled) {
+          val keyTypes = mapWithPipeline(conn, keys) { (pipeline, key) =>
+            pipeline.`type`(key)
+          }
+          keys.zip(keyTypes).filter(_._2 == keyType).map(_._1)
+        } else {
+          keys
+        }
+      val pipelineValues = mapWithPipeline(conn, filteredKeys) { (pipeline, key) =>
         persistence.load(pipeline, key, requiredColumns)
       }
       keys.zip(pipelineValues).map { case (key, value) =>
```

Comment:
Should it be `filteredKeys` instead of `keys` here? Why it was not spotted with unit tests?

Reply:
agree, it's strange that the tests pass :)

Reply:
Could you please add a test that intercepts an exception if the flag is not set.

Reply:
sure